programs/quarter_board/communication.c

```diff
@@ -0,0 +1,170 @@
+thread_t *communicationp;
+
+message_data_t message;
+message_data_t *msgp = &message;
+
+uint8_t value_to_char(uint8_t character) {
+    if (character > 9) {
+       character = character - 10 + 'A';
+    } else {
+       character = character + '0';
+    }
+    return character;
+}
+
+
+uint8_t char_to_value(uint8_t character) {
+    if (character >= 'A') {
+        character = character + 10 - 'A';
+    } else {
+        character = character - '0';
+    }
+    return character;
+}
+
+THD_WORKING_AREA(waSerialThread, 128);
+THD_FUNCTION(SerialThread, msgThread) {
+    static uint16_t command;
+    static uint8_t my_address;
+    static uint8_t quantity_of_slaves;
+    static uint8_t state = COMMAND_CHAR_0_READ;
+    uint8_t char_counter = 0;
+    uint8_t data_length = 0;
+    uint8_t i;
+    uint8_t received_char;
+    uint8_t rw = 0;
+
+    uint8_t *data_buffer = NULL;
+    uint8_t *operation_buffer = NULL;
+
+    while (1) {
+        received_char = sdGet(&SD1);
+
+        if ((received_char >> 7) & 1) {
+            if ((received_char >> 6) & 1) {
+                quantity_of_slaves = (received_char + my_address) & ~0xC0;
+                
+                if (operation_buffer == NULL) {
+                    operation_buffer = chCoreAlloc(sizeof(uint8_t) * quantity_of_slaves *
+                                                   MAX_DATA_CHAR_LENGHT);
+                }
+
+                if (data_buffer == NULL) {
+                    data_buffer = chCoreAlloc(sizeof(uint8_t) * quantity_of_slaves * 
+                                              MAX_DATA_CHAR_LENGHT);
+                }
+
+                received_char--;
+                sdPut(&SD1, received_char);
+            } else {
+                my_address = received_char & ~0x80;
+                received_char++;
+                sdPut(&SD1, received_char);
+            }
+        } else if (received_char == 0x0D || received_char == 0x0A) {
+            state = COMMAND_CHAR_0_READ; // communication reset - \r \n
+            sdPut(&SD1, received_char);
+            char_counter = 0;
+            command = 0;
+        } else if ((received_char >= '0' && received_char <= '9') || 
+                   (received_char >= 'A' && received_char <= 'F')) {
+            
+            received_char = char_to_value(received_char);
+
+            if (state <= COMMAND_END) {
+                command |= (received_char << (state * CHAR_BIT_LENGHT));
+
+                sdPut(&SD1, value_to_char(received_char));
+
+                state++;
+
+                if (state == COMMAND_END + 1) {
+                    //commmand_received
+                    for (i = 0; i < (quantity_of_slaves * MAX_DATA_CHAR_LENGHT); i++) {
+                        operation_buffer[i] = NO_OPERATION;
+                        data_buffer[i] = 0;
+                    }
+
+                    data_length = (command >> (OPERATION_BIT_LENGHT + RW_BIT));
+                    rw = (((command >> OPERATION_BIT_LENGHT) & 1) + 1);

```

Comment:
The outermost parentheses here are not required (same for the line above this one as well). Please remove them.